### DISCUSSION TITLE
How to reset the options of a Select widget

### DISCUSSION THREAD

**Discussion by PlasmaHH** [Textualize/textual#5305] [Q&A]
I tried to naively just use set_options to set them as that seemed to be the right thing:

```python
import textual.widgets

sel = textual.widgets.Select( [ ( "", "" ) ], classes = "select", id = "select" )
sel.set_options( [ ("A", "B") ] )
```

but that resulted in

```
> ./selecte.py 
Traceback (most recent call last):
  File "/home/plasmahh/src/python/./selecte.py", line 7, in <module>
    sel.set_options( [ ("A", "B") ] )
  File "/home/plasmahh/.venv/dev/lib64/python3.11/site-packages/textual/widgets/_select.py", line 452, in set_options
    self._setup_options_renderables()
  File "/home/plasmahh/.venv/dev/lib64/python3.11/site-packages/textual/widgets/_select.py", line 426, in _setup_options_renderables
    option_list = self.query_one(SelectOverlay)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/plasmahh/.venv/dev/lib64/python3.11/site-packages/textual/dom.py", line 1472, in query_one
    raise NoMatches(f"No nodes match {selector!r} on {self!r}")
textual.css.query.NoMatches: No nodes match <class 'textual.widgets._select.SelectOverlay'> on Select(id='select')
```

How can I update the options (without having to always create a new Select object)?

Bonus points: how to set the selected value right away without triggering an event?

**Comment by darrenburns** [comment]
You can't interact with widgets like this outside of a Textual app.

You'll need to move this code into a `compose` method in an `App`.

**Reply by PlasmaHH** [reply]
So everytime I want to change something in the select widget I have to re-compose my whole app? What is the use-case of the set_options() function then? In the compose you would just create a new Select() object anyways each time...

(My use-case is actually right  now in the ctor of my app, this was just a minimal example to showcase the problem)

**Reply by willmcgugan** [reply]
> So everytime I want to change something in the select widget I have to re-compose my whole app?

Not sure how you jumped to that conclusion, but no.

Widgets need to be in the DOM before you modify them. The earliest point you can work with a widget is in its Mount event.

Bonus points: https://textual.textualize.io/guide/events/#preventing-messages

**Reply by darrenburns** [reply]
@PlasmaHH Minimal examples really need to be something that we can run, so we can see context. The location of this code matters, as Will said.

You'll need to `yield Select(...)` inside your `compose`, and then `set_options` should work inside event handlers, etc.

**Reply by PlasmaHH** [reply]
> > So everytime I want to change something in the select widget I have to re-compose my whole app?
> 
> Not sure how you jumped to that conclusion, but no.

It was the 

> You'll need to move this code into a compose method in an App.

part, which sounds like it means that the code (more specifically the set_options) must run in the compose function.

So doing it in some code that is run via call_next/call_later is fine then?

(Minimal example from my pov is the smallest that exhibits the behaviour. And as I was not aware of that it must be run within certain parts of the program, that was exactly the point, had I known that I was not supposed to run it in __init__, it would have worked )

**Reply by TomJGooding** [reply]
Here's a simple example just to demonstrate, I'm not sure on what event you want to reset the options.

```python
from textual.app import App, ComposeResult
from textual.widgets import Footer, Select


class ExampleApp(App):
    BINDINGS = [("s", "set_options", "Set options")]

    def compose(self) -> ComposeResult:
        yield Select([("Before", "Before")])
        yield Footer()

    def action_set_options(self) -> None:
        select = self.query_one(Select)
        select.set_options([("After", "After")])
        self.notify("Select options set")


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Reply by darrenburns** [reply]
> Minimal example from my pov is the smallest that exhibits the behaviour.

Your example doesn't exhibit any behaviour because it doesn't run. It's just really hard (prone to misunderstandings, and time consuming for us and you) to offer advice without context. See https://github.com/Textualize/textual/discussions/3361 for more info.

The reason I made my initial comment is because in your example it looks like you've just created a Python script, imported `textual.widgets`, and added a couple of lines of code. We get people of all different experience levels asking questions, and some people do things like this.

> So doing it in some code that is run via call_next/call_later is fine then?

It's not directly related to `call_next`/`call_later`, it's more dependent on where you're running it from. Without context it's hard to say e.g. - Are you loading the options from and API on startup (if so, just `set_options` in `on_mount`)? Are you clicking a button somewhere and that's changing the options (`set_options` goes in your event handler)?

**Reply by PlasmaHH** [reply]
> > Minimal example from my pov is the smallest that exhibits the behaviour.
> 
> Your example doesn't exhibit any behaviour because it doesn't run. 

Well exactly that was the point, my original code didn't run and exhibited the same behaviour, so I stripped it down until the shown code was the minimum necessary to exhibit the same error. If questions here are meant to me only about code that is running fine I will try to take that into account for the next time.

> > So doing it in some code that is run via call_next/call_later is fine then?
> 
> It's not directly related to `call_next`/`call_later`, it's more dependent on where you're running it from. Without context it's hard to say e.g. - Are you loading the options from and API on startup (if so, just `set_options` in `on_mount`)? Are you clicking a button somewhere and that's changing the options (`set_options` goes in your event handler)?

Well the context where I want to set the options can be at a lot of places. I am updating one of my objects and this can cause the select options to change. And setting it can happen from a button, from a function called in via call_next or call_later (I have some functionality that uses this to "schedule" itself again as soon as it is done), or, and that was the most problematic part, somewhere in the applications setup where I prepare the widgets, which was initially in __init__ (and now for that widget I leave it empty and let it be populated in the first run from a call_next or call_later path )

**Reply by darrenburns** [reply]
> (and now for that widget I leave it empty and let it be populated in the first run from a call_next or call_later path )

You probably want to move this stuff into `on_mount` and remove the `call_next` and `call_later`. Doing that would be more idiomatic and also means that widgets are ready to work with (so `set_mount` will work).

**Reply by PlasmaHH** [reply]
> You probably want to move this stuff into `on_mount` and remove the `call_next` and `call_later`. Doing that would be more idiomatic and also means that widgets are ready to work with (so `set_mount` will work).

You mean for the initial setup ? Or not use call_next/later at all? If so, what would be a way to update some display not based on events but as fast as possible?

**Reply by darrenburns** [reply]
I mean you should do initial setup in `on_mount` without calling `call_next`/`call_later`.

`call_next` and `call_later` have their uses but I can't see why they'd be required based on what you've described.

> what would be a way to update some display not based on events but as fast as possible?

`on_mount` is the earliest point in time you can reliably update widgets.

**Reply by PlasmaHH** [reply]
> `on_mount` is the earliest point in time you can reliably update widgets.

Ok, that sounds like a good rule to work by

regarding the call_next/later thing, this seemed to be the easiest way I could find in the API to update all the data I display on the screen in the fastest way possible: get called, retrieve data, update widgets, schedule me to get called again.
